enzy_htp/analysis/residue_pka.py

```diff
@@ -0,0 +1,130 @@
+"""Submodule contains code for calculations the pKa of residues in the enzyme.
++ residue_pka()
+    calculate the pKa of residues in a protein from a Structure or an ensemble
+    of Structure()s
+
+Author: QZ Shao <shaoqz@icloud.com>
+Author: Robbie Ge
+
+Date: 2025-08-07
+"""
+from typing import List, Dict, Union, Callable, Tuple
+
+from enzy_htp import interface, config as eh_config
+from enzy_htp.structure import Structure, StructureEnsemble, Residue
+from enzy_htp.core import _LOGGER
+from enzy_htp.preparation.clean import remove_solvent
+
+
+def _filter_target_residues(
+        all_pka: Dict[Tuple[str, int], float], 
+        target_residues: Union[List[str], List[Residue], None]
+    ) -> Dict[Tuple[str, int], float]:
+    """Helper function to filter pKa results for target residues using chain_id + res_num for robust alignment."""
+    if target_residues is None:
+        return all_pka
+    
+    # Convert target_residues to (chain_id, res_num) tuples
+    target_keys = []
+    for res in target_residues:
+        if isinstance(res, str):
+            # Format: "A.100" (chain.residue_number)
+            if "." in res:
+                parts = res.split(".")
+                if len(parts) == 2:
```

Comment:
For something like this, it will improve readability to make a helper function something along the lines of `valid_res_code()` that will just check this. The level of indentation and nesting in this function hurts its readability.